wal/src/message_queue_impl/region_context.rs

```diff
@@ -0,0 +1,660 @@
+                delta.sequence_num + 1,
+                table_meta.next_sequence_num,
+            ) });
+        table_meta.next_sequence_num = delta.sequence_num + 1;
+
+        ensure!(table_meta.current_high_watermark < delta.offset + 1, Build { msg: format!("apply delta failed, 
+                high watermark in delta should't be less than or equal to the one in builder, but now are:{} and {}",
+                delta.offset + 1,
+                table_meta.current_high_watermark,
+            ) });
+        table_meta.current_high_watermark = delta.offset + 1;
+
+        table_meta
+            .start_sequence_offset_mapping
+            .insert(delta.sequence_num, delta.offset);
+
+        Ok(())
+    }
+
+    pub fn build(self) -> RegionContext {
+        debug!(
+            "Region meta data before building, region meta data:{:?}",
+            self.table_metas
+        );
+
+        let table_metas = self
+            .table_metas
+            .into_iter()
+            .map(|(table_id, table_meta_inner)| {
+                let table_meta = TableMeta {
+                    table_id,
+                    inner: Mutex::new(table_meta_inner),
+                };
+
+                (table_id, TableContext::new(table_meta))
+            })
+            .collect();
+
+        RegionContext {
+            inner: RwLock::new(RegionContextInner {
+                table_contexts: table_metas,
+            }),
+            region_id: self.region_id,
+        }
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+pub struct RegionMetaDelta {
+    table_id: TableId,
+    sequence_num: SequenceNumber,
+    offset: Offset,
+}
+
+#[allow(unused)]
+impl RegionMetaDelta {
+    pub fn new(table_id: TableId, sequence_num: SequenceNumber, offset: Offset) -> Self {
+        Self {
+            table_id,
+            sequence_num,
+            offset,
+        }
+    }
+}
+
+impl From<TableMetaData> for TableMetaInner {
+    fn from(table_meta_data: TableMetaData) -> Self {
+        let mut start_sequence_offset_mapping = BTreeMap::new();
+        if let Some(safe_delete_offset) = &table_meta_data.safe_delete_offset {
+            start_sequence_offset_mapping
+                .insert(table_meta_data.latest_marked_deleted, *safe_delete_offset);
+        }
+
+        TableMetaInner {
+            next_sequence_num: table_meta_data.next_sequence_num,
+            latest_marked_deleted: table_meta_data.latest_marked_deleted,
+            current_high_watermark: table_meta_data.current_high_watermark,
+            start_sequence_offset_mapping,
+        }
+    }
+}
+
+/// Table's logs writer.
+///
+/// NOTICE: Will lock it before writing to make writing to a specific table
+/// sequentially.
+#[derive(Debug)]
+pub struct TableWriter;
+
+impl TableWriter {
+    async fn write<M: MessageQueue>(
+        &self,
+        table_meta: &TableMeta,
+        _ctx: &manager::WriteContext,
```

Comment:
Why not remove this parameter if the `ctx` is not needed.

Reply:
I plan to add necessary `timeout` to every operation later.